src/translation-api.js

```diff
@@ -0,0 +1,107 @@
+import { and, eq } from 'drizzle-orm'
+import { kCreateWithDocId, kSelect } from './datatype/index.js'
+import { hashObject } from './utils.js'
+
+export const ktranslatedLanguageCodeToSchemaNames = Symbol(
+  'translatedLanguageCodeToSchemaNames'
+)
+export default class TranslationApi {
+  /** @type {Map<
+   * import('@mapeo/schema').TranslationValue['languageCode'],
+   * Set<import('@mapeo/schema/dist/types.js').SchemaName>>} */
+  #translatedLanguageCodeToSchemaNames = new Map()
+
+  /**
+   * @param {Object} opts
+   * @param {import('./datatype/index.js').DataType<
+   *   import('./datastore/index.js').DataStore<'config'>,
+   *   typeof import('./schema/project.js').translationTable,
+   *   'translation',
+   *   import('@mapeo/schema').Translation,
+   *   import('@mapeo/schema').TranslationValue
+   * >}  opts.dataType
+   * @param {typeof import('./schema/project.js').translationTable} opts.table
+   */
+  constructor({ dataType, table }) {
+    this.dataType = dataType
+    this.table = table
+  }
+
+  /**
+   * @param {import('@mapeo/schema').TranslationValue} value
+   */
+  async put(value) {
+    /* eslint-disable no-unused-vars */
+    const { message, ...identifiers } = value
+    const docId = hashObject(identifiers)
+    let existing
+    try {
+      existing = await this.dataType.getByDocId(docId)
+      if (existing) {
+        await this.dataType.update(existing.versionId, value)
+      }
+    } catch (e) {
+      existing = await this.dataType[kCreateWithDocId](docId, value)
+    }
+    return existing.docId
+  }
+
+  /**
+   * @param {import('type-fest').SetOptional<
+   * Omit<import('@mapeo/schema').TranslationValue,'schemaName' | 'message'>,
+   * 'fieldRef' | 'regionCode'>} value - We omit the schemaName and message, and we make fieldRef and RegionCode optional
```

Comment:
Nitpick: I think it's clear without this comment, and this comment isn't really for consumers of this function. I'd remove.

```suggestion
   * 'fieldRef' | 'regionCode'>} value
```